designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,99 @@
+# INuGetProjectServices
+
+* Status: In Review
+* Author(s): [Andy Zivkovic](https://github.com/zivkan)
+* Issue: [1](https://github.com/NuGet/Home/issues/9577) IVsPackageInstallerServices should be async
+
+## Problem Background
+
+The existing IVsPackageInstallerServices service is synchronous, and is being called by several components on the UI thread, causing UI delays in Visual Studio. We are creating new async APIs that will allow us to unblock the UI thread, and use the Visual Studio Service Broker to work in Visual Studio CodeSpaces (and maybe Live Share).
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Quickly return incomplete task to unblock thread.
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+public interface INuGetProjectServices
+{
+    /// <summary>Gets the list of packages installed in a project.</summmry>
+    /// <param name="project">Path and filename to project file.</param>
+    /// <param name="options">Options to use. If null, uses GetInstalledPakcagesOptions.Default.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// If the project does not exist in the solution, throws ArgumentException.
+    /// If the project is not ready, throws NuGetProjectNotReadyException.
+    /// </return>
+    Task<IReadOnlyList<NuGetPackage>> GetInstalledPackagesAsync(string project, GetInstalledPackagesOptions options, CancellationToken cancellationToken);
+}
+
+public sealed class NuGetPackage
+{
+    /// <summary>The package id.</summary>
+    string Id { get; }
+
+    /// <summary>The package version</summary>
+    /// <remarks>
+    /// If the project uses packages.config, this will be the installed package version.
+    /// If the project uses PackageReference, this would ideally be the resolved package version, but may be the requested package version.
+    /// </remarks>
+    string Version { get; }
+
+    internal NuGetPackage(string id, string version);
+}
+
+public sealed class GetInstalledPackagesOptions
+{
+    public static GetInstalledPackageOptions Default { get; }
+
+    // in the future may include options such as WaitForNomination, WaitForRestore.
+}
+
+public sealed class NuGetProjectNotReadyException : InvalidOperationException
```

Comment:
I'm not sure we'd want to extend from InvalidOperationException. This implies a user error. That's not necessarily true. 
Maybe we tell customers they need to couple this with an operation progress call to check if the project is ready.

Reply:
My intention was to be compatible with `IVsPackageInstallerServices.GetInstalledPackages`. But since this is a new API not intended to be compatible, then you're right, I don't think it makes sense.